server/server_test.go

```diff
@@ -479,6 +479,59 @@ var _ = Describe("Server package test", func() {
 			testURL("DELETE", serverPluralURL+"/"+serverID, memberTokenID, nil, http.StatusUnauthorized)
 			testURL("DELETE", serverPluralURL+"/"+serverID, adminTokenID, nil, http.StatusNoContent)
 		})
+		It("should work with filter based condition", func() {
+			// Tests query: `SELECT ... WHERE tenant_id.. OR (state = UP AND level IN (2,3)
+			expectedToContainTest := func(expectedCount int, res interface{}) {
+				Expect(res.(map[string]interface{})["filter_tests"]).To(HaveLen(expectedCount))
+			}
+			testPluralURL := baseURL + "/v2.0/filter_tests"
+			testUp := map[string]interface{}{
+				"state": "UP",
+				"level": 2,
+			}
+			testUpLevel3 := map[string]interface{}{
+				"state": "UP",
+				"level": 3,
+			}
+			testUpLevel0ID := "testUP"
+			testUpLevel0 := map[string]interface{}{
+				"id":    testUpLevel0ID,
+				"state": "UP",
+				"level": 0,
+			}
+			testDownID := "testDOWN"
+			testDown := map[string]interface{}{
+				"id":    testDownID,
+				"state": "DOWN",
+				"level": 2,
+			}
+			testUpdate := map[string]interface{}{
+				"state": "UP",
+				"level": 2,
+			}
+
+			var res interface{}
+			testURL("POST", testPluralURL, memberTokenID, testUp, http.StatusCreated)
+			testURL("POST", testPluralURL, memberTokenID, testUpLevel3, http.StatusCreated)
+			testURL("POST", testPluralURL, memberTokenID, testUpLevel0, http.StatusCreated)
+			testURL("POST", testPluralURL, memberTokenID, testDown, http.StatusCreated)
+			res = testURL("GET", testPluralURL, memberTokenID, nil, http.StatusOK)
+			expectedToContainTest(4, res)
+			res = testURL("GET", testPluralURL, powerUserTokenID, nil, http.StatusOK)
+			expectedToContainTest(2, res)
+
+			testURL("PUT", testPluralURL+"/"+testDownID, memberTokenID, testUpdate, http.StatusOK)
+			res = testURL("GET", testPluralURL, memberTokenID, nil, http.StatusOK)
+			expectedToContainTest(4, res)
+			res = testURL("GET", testPluralURL, powerUserTokenID, nil, http.StatusOK)
+			expectedToContainTest(3, res)
+
+			testURL("PUT", testPluralURL+"/"+testUpLevel0ID, memberTokenID, testUpdate, http.StatusOK)
+			res = testURL("GET", testPluralURL, memberTokenID, nil, http.StatusOK)
+			expectedToContainTest(4, res)
+			res = testURL("GET", testPluralURL, powerUserTokenID, nil, http.StatusOK)
+			expectedToContainTest(4, res)
+		})
```

Comment:
This is testing the scenario "tenant A resources can be read tenant A and other tenants if conditions match".

Can current implementation cover another scenario "some tenant A resources can be read by tenant A but cannot be updated or deleted even by tenant A (owner) if some conditions match"? This is admin-managed user resources scenario.

Reply:
Good idea - added. I've also added `not equal` operator.
For more operators, we have to change our squirrel vendor to https://github.com/Masterminds/squirrel - `lann`'s project is deprecated.